bitcoin/src/blockdata/script/borrowed.rs

```diff
@@ -167,6 +167,9 @@ impl Script {
     /// Returns witness version of the script, if any, assuming the script is a `scriptPubkey`.
     #[inline]
     pub fn witness_version(&self) -> Option<WitnessVersion> {
+        if !self.is_witness_program() {
```

Comment:
We can convert this witness_version into a private helper function `witness_version_unchecked` and make a public `witness_version` method. 
The already existing calls to `witness_version` in `is_p2wsh` don't need this check. 

I am also okay with the current solution, and don't mind slight efficiency loss due to redundant checks.

Reply:
I think we can do a slightly simpler thing. We can move the logic from `is_witness_program` into `witness_version`, replacing the `WitnessVersion::try_from(ver_opcode).is_ok()` with a line that returns the result of `try_from` without calling `is_ok`. (I guess, calling `ok()` to turn the result into an option.)

Then `is_witness_program`'s implementation can just be `self.witness_version().is_some()`.

Reply:
In any case, I don't really like this PR's redundant logic. Not because it's expensive but because it's hard to confirm that the two functions are consistent.

Reply:
Fair and fair - will re-spin. Thanks

Reply:
Much better